### PR TITLE
Added support for scanning 7-segment displays on FPGA-boards (issue #1764)

### PR DESCRIPTION
**Pull request by BFH-ktt1**

### CODE REVIEW

File: src/main/java/com/cburch/logisim/fpga/data/MapComponent.java [hunk 1]
```diff
@@ -191,7 +191,8 @@ public boolean isMapped(int pin) {
   public boolean isInternalMapped(int pin) {
     if (pin < 0 || pin >= nrOfPins) return false;
     return isBoardMapped(pin)
-        && maps.get(pin).getIoComp().getType().equals(IoComponentTypes.LedArray);
+        && (maps.get(pin).getIoComp().getType().equals(IoComponentTypes.LedArray) 
+        || maps.get(pin).getIoComp().getType().equals(IoComponentTypes.SevenSegmentScanning));
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/data/MapComponent.java:195]: Check the logic here. You are calling maps.get(pin) in the right side of the 'or' when isBoardMapped(pin) is false. Is that what you intend to do?

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java [hunk 2]
```diff
@@ -91,6 +94,47 @@ private static void buildPinTable(
...
+    NrOfPins.put(INPUT_ID, 0);
+    NrOfPins.put(IO_ID, 0);
+    var nrOfPins = 8;
+    pinLabels.addAll(SevenSegment.getLabels());
+    final var nrOfControlPins = (driveMode == SevenSegmentScanningDriving.SEVEN_SEG_DECODED) 
+        ? Math.max((int) Math.ceil(Math.log(nrOfDigits) / Math.log(2.0)), nrOfDecodedBits) : nrOfDigits;
+    nrOfPins += nrOfControlPins;
+    final var pinName = (driveMode == SevenSegmentScanningDriving.SEVEN_SEG_DECODED) ? "A" : "Seg";
+    for (var contrPin = 0; contrPin < nrOfControlPins; contrPin++)
+      pinLabels.add(String.format("%s%d", pinName, contrPin));
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:128]: Use curly brackets on this for loop.

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java [hunk 3]
```diff
@@ -426,6 +491,53 @@ else if (cnt < nrOfPins.get(INPUT_ID) + nrOfPins.get(OUTPUT_ID))
...
     }
+    if (myType.equals(IoComponentTypes.SevenSegmentScanning)) {
+      final var panel = new JPanel();
+      panel.setLayout(new GridBagLayout());
+      panel.setBorder(BorderFactory.createTitledBorder(
+              BorderFactory.createLineBorder(Color.BLACK, 2, true), S.get("FpgaScanningDefinition")));
+      scanningPanel.setLayout(new GridBagLayout());
+      rowSize.removeAll();
+      for (var nrOfSegments = 2; nrOfSegments < 16; nrOfSegments++) 
+        rowSize.addItem(nrOfSegments);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:502]: Use curly brackets.

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java [hunk 4]
```diff
@@ -426,6 +491,53 @@ else if (cnt < nrOfPins.get(INPUT_ID) + nrOfPins.get(OUTPUT_ID))
...
+      rowSize.setSelectedIndex(info.getNrOfRows() - 2);
+      eEncoding.removeAll();
+      for (var val : SevenSegmentScanningDriving.getDisplayStrings())
+        eEncoding.addItem(val);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:506]: Use curly brackets.

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java [hunk 5]
```diff
@@ -426,6 +491,53 @@ else if (cnt < nrOfPins.get(INPUT_ID) + nrOfPins.get(OUTPUT_ID))
...
+      eEncoding.setSelectedIndex(info.getArrayDriveMode());
+      decodedBits = new JComboBox<Integer>();
+      for (var nrOfDecodedBits = 1; nrOfDecodedBits < 6; nrOfDecodedBits++)
+          decodedBits.addItem(nrOfDecodedBits);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:510]: Use curly brackets and fix indentation.

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java [hunk 6]
```diff
@@ -426,6 +491,53 @@ else if (cnt < nrOfPins.get(INPUT_ID) + nrOfPins.get(OUTPUT_ID))
...
+      if (info.getArrayDriveMode() != SevenSegmentScanningDriving.SEVEN_SEG_DECODED)
+          decodedBits.setVisible(false);
+      else
+          decodedBits.setSelectedIndex(info.getNrOfColumns() - 1);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:514]: Use curly brackets and fix indentation.

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java [hunk 7]
```diff
@@ -157,7 +172,22 @@ public ToplevelHdlGeneratorFactory(
               ledArray.getNrOfRows(),
               ledArray.getNrOfColumns(),
               myLedArrays.indexOf(ledArray));
-      for (final var port : ports.keySet()) myPorts.add(Port.OUTPUT, port, ports.get(port), null);
+      for (final var port : ports.keySet()) 
+          myPorts.add(Port.OUTPUT, port, ports.get(port), null);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:176]: Use curly brackets.

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java [hunk 8]
```diff
@@ -157,7 +172,22 @@ public ToplevelHdlGeneratorFactory(
...
+              scanningSevenSeg.getNrOfRows(), 
+              myScanningSevenSegs.indexOf(scanningSevenSeg)));
+      final var ports = 
+          SevenSegmentScanningGenericHdlGenerator.getExternalSignals(
+              scanningSevenSeg.getArrayDriveMode(),
+              scanningSevenSeg.getNrOfRows(),
+              scanningSevenSeg.getNrOfColumns(),
+              myScanningSevenSegs.indexOf(scanningSevenSeg));
+      for (final var port : ports.keySet())
+        myPorts.add(Port.OUTPUT, port, ports.get(port), null);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:190]: Use curly brackets.

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java [hunk 9]
```diff
@@ -219,6 +256,14 @@ public LineBuffer getComponentDeclarationSection(Netlist theNetlist, AttributeSe
           components.add(worker.getComponentInstantiation(theNetlist, null, name)).empty();
       }
     }
+    for (final var type : SevenSegmentScanningDriving.DRIVING_STRINGS) {
+      if (hasScanningSevenSegmentType(type)) {
+        final var worker = SevenSegmentScanningGenericHdlGenerator.getSpecificHDLGenerator(type);
+        final var name = SevenSegmentScanningGenericHdlGenerator.getSpecificHDLName(type);
+        if (worker != null && name != null)
+          components.add(worker.getComponentInstantiation(theNetlist, null, name)).empty();
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:264]: Use curly brackets.

File: src/main/java/com/cburch/logisim/std/io/SevenSegmendScanningSelectedHdlGenerator.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    super();
+    myParametersList.add(SELECT_DIGIT_INVERTED_STRING, SELECT_DIGIT_INVERTED_ID);
+    myPorts.removePorts();
+    myPorts
+        .add(Port.INPUT, SevenSegmentScanningGenericHdlGenerator.SevenSegmentSegmenInputs, NR_OF_SEGMENTS_ID, 9)
+        .add(Port.OUTPUT, SevenSegmentScanningGenericHdlGenerator.SevenSegmentControlOutput, NR_OF_DIGITS_ID, 10)
+        .add(Port.INPUT, TickComponentHdlGeneratorFactory.FPGA_CLOCK, 1, 11);
+    var id = 0;
+    for (final var segName : SevenSegment.getLabels())
+      myPorts.add(Port.OUTPUT, segName, 1, id++);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/std/io/SevenSegmendScanningSelectedHdlGenerator.java:39]: Use curly brackets.

File: src/main/java/com/cburch/logisim/std/io/SevenSegmentScanningDecodedHdlGeneratorFactory.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+  }
+    
+  public static LineBuffer getPortMap(int id) {
+    final var ports = new HashMap<String, String>();
+    ports.put(SevenSegmentScanningGenericHdlGenerator.SevenSegmentSegmenInputs,
+            String.format("s_%s%d", SevenSegmentScanningGenericHdlGenerator.InternalSignalName, id));
+    ports.put(TickComponentHdlGeneratorFactory.FPGA_CLOCK, TickComponentHdlGeneratorFactory.FPGA_CLOCK);
+    ports.put(SevenSegmentScanningGenericHdlGenerator.SevenSegmentControlOutput, String.format("Displ%dSelect", id));
+    for (final var segName : SevenSegment.getLabels())
+        ports.put(segName, String.format("Displ%d_%s", id, segName));
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/std/io/SevenSegmentScanningDecodedHdlGeneratorFactory.java:89]: Use curly brackets.

File: src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java [hunk 12]
```diff
@@ -313,6 +329,14 @@ public FpgaIoInformationContainer(Node DocumentInfo) {
       for (var i = 0; i < nrOfPins; i++)
         myOutputPins.add(i);
     }
+    if (myType.equals(IoComponentTypes.SevenSegmentScanning)) {
+      nrOfExternalPins = nrOfPins;
+      nrOfPins = nrOfRows * 8;
+      setNrOfPins(nrOfPins);
+      myOutputPins.clear();
+      for (var pinNr = 0; pinNr < nrOfPins; pinNr++)
+        myOutputPins.add(pinNr);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java:338]: Use curly brackets here.

File: src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java [hunk 13]
```diff
@@ -1010,6 +1065,8 @@ public boolean tryMap(JPanel parent) {
     if (selComp == null) return false;
     if (myType.equals(IoComponentTypes.LedArray))
       return tryLedArrayMap(parent);
+    if (myType.equals(IoComponentTypes.SevenSegmentScanning))
+      return tryScanningMap(parent);
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java:1069]: Use curly brackets.

**Review by davidhutchens — COMMENTED** [review]
I made some comments about style and one about a potential logic error.

File: src/main/java/com/cburch/logisim/fpga/data/MapComponent.java (hunk 1 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/data/MapComponent.java:195]: I think the confusion is wrong identation, the `or` is inside brackets. The brackets with an and with isBoardMapped(pin). Let me know if this is the case

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java (hunk 2 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:128]: Done

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java (hunk 3 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:502]: done

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java (hunk 4 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:506]: done

File: src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java (hunk 5 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/gui/FpgaIoInformationSettingsDialog.java:510]: done

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java (hunk 7 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:176]: done

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java (hunk 8 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:190]: done

File: src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java (hunk 9 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/hdlgenerator/ToplevelHdlGeneratorFactory.java:264]: done

File: src/main/java/com/cburch/logisim/std/io/SevenSegmendScanningSelectedHdlGenerator.java (hunk 10 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/std/io/SevenSegmendScanningSelectedHdlGenerator.java:39]: done

File: src/main/java/com/cburch/logisim/std/io/SevenSegmentScanningDecodedHdlGeneratorFactory.java (hunk 11 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/std/io/SevenSegmentScanningDecodedHdlGeneratorFactory.java:89]: done

File: src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java (hunk 12 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java:338]: done

File: src/main/java/com/cburch/logisim/fpga/data/MapComponent.java (hunk 1 above)
Comment (davidhutchens) [reply src/main/java/com/cburch/logisim/fpga/data/MapComponent.java:195]: Ah. You are correct. Indenting the || line might help.

File: src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java (hunk 13 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/data/FpgaIoInformationContainer.java:1069]: done

File: src/main/java/com/cburch/logisim/fpga/data/MapComponent.java (hunk 1 above)
Comment (BFH-ktt1) [reply src/main/java/com/cburch/logisim/fpga/data/MapComponent.java:195]: Thanks! Done

**Review by davidhutchens — APPROVED** [review]
Looks good to me.

**Pull request merged**